File: backtest-terminal-go/stocks.go
package main

import (
	"sort"
	"strings"
	"sync"
)

// Asset cache shared between the autocomplete func (called inline on each
// keystroke) and the loadAssets goroutine that fills it. RWMutex because
// reads vastly outnumber writes — the catalogue is fetched once and then
// queried per keystroke.
var (
	assetMu      sync.RWMutex
	assetSymbols []string          // sorted, for binary-search prefix lookup
	assetNames   map[string]string // symbol -> display name
)

// loadAssets pulls all active, tradable US equity symbols from Alpaca and
// caches them. Fire-and-forget from a goroutine at startup — autocomplete
// just returns no matches until the fetch completes (~1 s on a normal
// connection). Returns silently on error so we don't block the UI.
func loadAssets() {
	if client == nil {
		return
	}
	assets, err := client.GetAssets()
	if err != nil {
		return
	}

	syms := make([]string, 0, len(assets))
	names := make(map[string]string, len(assets))
	for _, a := range assets {
		if a.Tradable {
			syms = append(syms, a.Symbol)
			names[a.Symbol] = a.Name
		}
	}
	sort.Strings(syms)

	assetMu.Lock()
	assetSymbols = syms
	assetNames = names
	assetMu.Unlock()
}

// filterStocks returns up to `limit` autocomplete entries matching prefix.
// Fast path: binary-search ticker-prefix matches. Slow path: linear scan of
// company-name substrings so typing "apple" surfaces AAPL.
//
// Each returned entry is formatted "SYMBOL  Company Name". Callers should
// strip on the first whitespace to recover the bare ticker.
func filterStocks(prefix string, limit int) []string {
	prefix = strings.ToUpper(prefix)

	assetMu.RLock()
	defer assetMu.RUnlock()

	if len(assetSymbols) == 0 {
		return nil
	}

	seen := make(map[string]bool)
	var out []string

	start := sort.SearchStrings(assetSymbols, prefix)
	for i := start; i < len(assetSymbols) && len(out) < limit; i++ {
		sym := assetSymbols[i]
		if !strings.HasPrefix(sym, prefix) {
			break
		}
		seen[sym] = true
		name := assetNames[sym]
		if len(name) > 38 {
			name = name[:35] + "…"
		}
		out = append(out, sym+"  "+name)
	}

	if len(out) < limit {
		lower := strings.ToLower(prefix)
		for _, sym := range assetSymbols {
			if len(out) >= limit {
				break
			}
			if seen[sym] {
				continue
			}
			name := assetNames[sym]
			if strings.Contains(strings.ToLower(name), lower) {
				seen[sym] = true
				display := name
				if len(display) > 38 {
					display = display[:35] + "…"
				}
				out = append(out, sym+"  "+display)
			}
		}
	}

	return out
}

File: chart-compare-gui-rust/src/stocks.rs
use std::collections::HashMap;
use std::sync::RwLock;

use crate::api::Asset;

pub struct AssetCache {
    symbols: RwLock<Vec<String>>,         // sorted, for binary-search prefix lookup
    names: RwLock<HashMap<String, String>>, // symbol -> company name
}

impl AssetCache {
    pub fn new() -> Self {
        AssetCache {
            symbols: RwLock::new(Vec::new()),
            names: RwLock::new(HashMap::new()),
        }
    }

    pub fn load(&self, assets: Vec<Asset>) {
        let mut syms: Vec<String> = Vec::with_capacity(assets.len());
        let mut names: HashMap<String, String> = HashMap::with_capacity(assets.len());
        for a in assets {
            if a.tradable {
                syms.push(a.symbol.clone());
                names.insert(a.symbol, a.name);
            }
        }
        syms.sort();
        *self.symbols.write().unwrap() = syms;
        *self.names.write().unwrap() = names;
    }

    pub fn company_name(&self, sym: &str) -> String {
        self.names
            .read()
            .unwrap()
            .get(sym)
            .cloned()
            .unwrap_or_default()
    }

    /// Returns up to `limit` "SYMBOL  Company Name" suggestions matching `prefix`.
    /// Fast path: binary-search ticker prefix. Slow path: substring of company name.
    pub fn filter(&self, prefix: &str, limit: usize) -> Vec<(String, String)> {
        let prefix = prefix.to_ascii_uppercase();
        let symbols = self.symbols.read().unwrap();
        let names = self.names.read().unwrap();
        if symbols.is_empty() || prefix.is_empty() {
            return Vec::new();
        }

        let mut out: Vec<(String, String)> = Vec::with_capacity(limit);
        let mut seen: std::collections::HashSet<&str> = std::collections::HashSet::new();

        let start = symbols.partition_point(|s| s.as_str() < prefix.as_str());
        for sym in &symbols[start..] {
            if !sym.starts_with(&prefix) || out.len() >= limit {
                break;
            }
            seen.insert(sym.as_str());
            let mut name = names.get(sym).cloned().unwrap_or_default();
            if name.chars().count() > 38 {
                name = format!("{}…", name.chars().take(35).collect::<String>());
            }
            out.push((sym.clone(), name));
        }

        if out.len() < limit {
            let lower = prefix.to_ascii_lowercase();
            for sym in symbols.iter() {
                if out.len() >= limit {
                    break;
                }
                if seen.contains(sym.as_str()) {
                    continue;
                }
                let name = names.get(sym).cloned().unwrap_or_default();
                if name.to_ascii_lowercase().contains(&lower) {
                    let display = if name.chars().count() > 38 {
                        format!("{}…", name.chars().take(35).collect::<String>())
                    } else {
                        name
                    };
                    out.push((sym.clone(), display));
                }
            }
        }
        out
    }
}
